### PR TITLE
refs #12771 \yii\web\User::can() and guest

### PR DESCRIPTION
**Pull request by pchapl**
| Q | A |
| --- | --- |
| Is bugfix? | no |
| New feature? | no |
| Breaks BC? | no |
| Tests pass? | yes |
| Fixed issues | #12771 |

Skip \yii\rbac\PhpManager::checkAccessRecursive and \yii\rbac\DbManager::checkAccessRecursive if role assignments are empty.
I checked code of `\yii\rbac\PhpManager::checkAccessRecursive` and `\yii\rbac\DbManager::checkAccessRecursive`, and I'm sure that result always will be "false" if `$assignments` is empty array and `\yii\rbac\BaseManager::$defaultRoles` is empty array too.
On the other hand, I'm not sure it matters.

### CODE REVIEW

File: framework/CHANGELOG.md [hunk 1]
```diff
@@ -94,6 +95,7 @@ Yii Framework 2 Change Log
 - Enh #12580: Make `yii.js` comply with strict and non-strict javascript mode to allow concatenation with external code (mikehaertl)
 - Enh #12664: Added support for wildcards for `optional` at `yii\filters\auth\AuthMethod` (mg-code)
 - Enh #12744: Added `afterInit` event to `yii.activeForm.js` (werew01f)
+- Enh #12499: When AJAX validation in enabled, `yii.activeForm.js` will run it forcefully on form submit to display all possible errors (silverfire)
```
Comment (SilverFire) [inline framework/CHANGELOG.md]: This line should be removed

File: framework/rbac/BaseManager.php [hunk 2]
```diff
@@ -220,4 +220,15 @@ protected function executeRule($user, $item, $params)
...
         }
     }
+
+    /**
+     * Check that there no any roles in user's role assignments and in default roles
+     *
+     * @param Assignment[] $assignments list of user's role assignments
+     * @return bool true if empty
+     */
+    protected function emptyRoles(array $assignments)
```
Comment (SilverFire) [inline framework/rbac/BaseManager.php]: We prefix this kind of methods with `has` or `is` to make it easier to understand, what does the method do.

Maybe, `hasNoAssignments()`?

File: framework/rbac/BaseManager.php [hunk 3]
```diff
@@ -220,4 +220,15 @@ protected function executeRule($user, $item, $params)
             throw new InvalidConfigException("Rule not found: {$item->ruleName}");
         }
     }
+
+    /**
+     * Check that there no any roles in user's role assignments and in default roles
```
Comment (SilverFire) [inline framework/rbac/BaseManager.php]: Method checks, whether $assignments array is empty and [[defaultRoles]] are empty as well.

File: framework/rbac/BaseManager.php [hunk 4]
```diff
@@ -220,4 +220,15 @@ protected function executeRule($user, $item, $params)
...
+     *
+     * @param Assignment[] $assignments list of user's role assignments
```
Comment (SilverFire) [inline framework/rbac/BaseManager.php]: `@param Assignment[] $assignments array of user's assignments`

File: framework/rbac/BaseManager.php [hunk 5]
```diff
@@ -220,4 +220,15 @@ protected function executeRule($user, $item, $params)
...
+     * @return bool true if empty
```
Comment (SilverFire) [inline framework/rbac/BaseManager.php]: `@return bool whether $assignments array is empty and [[defaultRoles]] are empty as well`

File: framework/rbac/BaseManager.php [hunk 6]
```diff
@@ -220,4 +220,15 @@ protected function executeRule($user, $item, $params)
...
+     */
+    protected function emptyRoles(array $assignments)
+    {
+        return count($assignments) === 0 && count($this->defaultRoles) === 0;
```
Comment (SilverFire) [inline framework/rbac/BaseManager.php]: Could be simplified to 

```php
return empty($assignments) && empty($this->defaultRoles);
```

File: framework/rbac/DbManager.php [hunk 7]
```diff
@@ -119,6 +119,12 @@ public function init()
     public function checkAccess($userId, $permissionName, $params = [])
     {
         $assignments = $this->getAssignments($userId);
+
+        // nothing to check, user has no any rights
```
Comment (SilverFire) [inline framework/rbac/DbManager.php]: Could be removed

File: framework/rbac/PhpManager.php [hunk 8]
```diff
@@ -97,6 +97,12 @@ public function init()
     public function checkAccess($userId, $permissionName, $params = [])
     {
         $assignments = $this->getAssignments($userId);
+
+        // nothing to check, user has no any rights
```
Comment (SilverFire) [inline framework/rbac/PhpManager.php]: Could be removed

**Review by SilverFire — CHANGES_REQUESTED** [review]

File: framework/CHANGELOG.md (hunk 1 above)
Comment (Ni-san) [inline framework/CHANGELOG.md]: done

File: framework/rbac/BaseManager.php (hunk 3 above)
Comment (Ni-san) [inline framework/rbac/BaseManager.php]: done

File: framework/rbac/BaseManager.php (hunk 4 above)
Comment (Ni-san) [inline framework/rbac/BaseManager.php]: done

File: framework/rbac/BaseManager.php (hunk 5 above)
Comment (Ni-san) [inline framework/rbac/BaseManager.php]: done

File: framework/rbac/BaseManager.php (hunk 2 above)
Comment (Ni-san) [inline framework/rbac/BaseManager.php]: done

File: framework/rbac/DbManager.php (hunk 7 above)
Comment (Ni-san) [inline framework/rbac/DbManager.php]: done

**Review by SilverFire — APPROVED** [review]
Looks good to me

**Review by dynasource — CHANGES_REQUESTED** [review]
Thanks for putting effort in improving Yii2.

A few things:
- there is duplication of logic in DbManager & PhpManager which should be prevented. I would expect a new function like 'getActiveAssignements' or something like that.
- inverse logic is applied in the method name 'hasNoAssignments'. This should be formulated positively like 'hasAssignments'
- I a still missing a clear explanation in the method docs why the added method is useful. It states what it does, but not why its useful.
- new methods should  get a @since tag

**Comment by SilverFire** [comment]
Sorry for this mess, it looks like a GitHub bug :(

**Comment by SilverFire** [comment]
A few words what happen to this PR: 

1. I've synchronized it with master and pushed changes to the branch in @Ni-san's fork. GitHub got crazy and showed a diff for this PR containing hundreds of commits.
2. I've removed all this mess and force-pushed clear branch
3. No changes suggested by @Ni-san survived in his branch after this force-push. Sorry.

**Comment by SilverFire** [comment]
Restored changes from `git reflog`, pushed to the master branch: https://github.com/yiisoft/yii2/commit/5795b3937eab3764a79e180f259dbabd6f655dad

**Comment by SilverFire** [comment]
The PR is merged. Thank yoou

**Pull request closed without merging**